File: 02/Conversion.c
#include <stdio.h>
#include <math.h>

// compile with gcc -o Conversionc ./Conversion.c -lm
int main(void) {

/* Declare variables */
   int i,inum,tmp,numdigits;
   float fnum;
   char binnum[60];

/* Intialise 4-byte integer */
//   inum = 33554431; //33554431 = 2^25-1 = 2^0 + 2^1 + ... + 2^24, needs 25 digits to represent accurately
//   inum = 16777215;//no rounding error
//   inum = 16777215 + 1; //no rounding error
   inum = 16777215 + 2; //rounding error
//   inum = 16777215 + 3; // no rounding error
/* Convert to 4-byte float */
   fnum = (float) inum; //after conversion, 33554431 becomes 33554432.0000, by rounding-off the rightmost digit '1'
   // if the distance between the two leftmost and rightmost digits '1' is greater than 23, rounding error occurs after conversion


/* Convert to binary number (string)*/
   i = 0; tmp = inum;
   while (tmp > 0) {
     sprintf(&binnum[i],"%1d",tmp%2);
     tmp = tmp/2;
     i++;
   }
   printf("number of digits converted = %d\n",i);

/* Terminate the string */
   binnum[i] = '\0';
   
/* Reverse the 0/1 string */
   int n = i-1;
   int im = floor(i/2.);
   char c;
   for(int j=0;j<im;j++){
      c = binnum[j];
      binnum[j]=binnum[n-j];
      binnum[n-j]=c;
   }

/* Complete the expression */
/* the calculation is not accurate: when inum=16777215 + 2, it gives numdigits=24 (no conversion error), but actually fnum=16777215+1;
   it cannot be used to judge if conversion error will occur or not.
*/
   numdigits = ceil(logf(fnum)/logf(2.) ); 
   printf("The number of digits is %d\n",numdigits);
   
   printf("inum=%d,  fnum=%f, inum in binary=%s\n",
      inum,fnum,binnum);
      
   //with 24 digits, the maximal number is identified by
   //2^0 + 2^1 + 2^2 + ... + 2^23 = (1-2^24)/(1-2)=2^24-1
//   int maxintval=ceil(pow(2.,24.));
//   maxintval=maxintval-1.;
//   printf("max integer is %d\n",maxintval); // the result is 16777215, with 25 digits the result is 2^25-1=33554431

}

File: 02/Conversion.f90
program conversion
   
   implicit none
   
! Declare variables 
   integer (kind=4) :: i,inum,tmp,numdigits
   integer :: binnum(60)
   real(kind=4) :: fnum
   integer (kind=4) :: n,im

! Intialise 4-byte integer
!   inum = 33554431
   inum = 16777215 + 2; !to match content in file Conversion.c
!   inum = 26 !used to verify the reversing of 0/1 in binnum with un-symmetric digits
! Convert to 4-byte float
   fnum = real(inum)       

! Binary converter using modulus   
   i=0
   tmp=inum
   do while (tmp > 0)
      i = i+1
      if(mod(tmp,2)==0) then
             binnum(i) = 0
        else
             binnum(i) = 1
      end if
      tmp = tmp/2
   end do
   !store the number of digits
   n = i
   
!   write(*,'(A,I5)')'number of digits converted =',i
   
   !output inum in binary form, stored in binnum
!   write(6,'(a)',advance='No') '(before reversing) inum in binary='
!   DO i=1,n
!      write(6,'(I1)',advance='No') binnum(i)
!   END DO
!   write(6,*)

! Reverse the 0/1 string
  im = n/2 !if n=5, then im=2
!  print*, 'n=',n,', im=',im
  DO i=1, im
    tmp = binnum(i)
    binnum(i) = binnum(n+1-i)
    binnum(n+1-i) = tmp
  end do
  
! TODO Complete the expression
   numdigits = nint(log(fnum)/log(2.)) 
   write(6,*) ' The number of digits is ',numdigits

   write(6,'(a,i0,a,f0.1,a,b0)')  'inum=',inum,', fnum=',fnum,', inum in binary=',inum
   
   !output inum in binary form, stored in binnum
   write(6,'(a)',advance='No') 'inum in binary='
   DO i=1,n
      write(6,'(I1)',advance='No') binnum(i)
   END DO
   write(6,*)


end program conversion
